### ISSUE TITLE
Support protocol version 2

### ISSUE THREAD

**Issue by joebandenburg** [joebandenburg/libaxolotl-javascript#3]
It would be good to be interoperable with older clients. Although it would be worth finding out if there are many version 2 clients out there before starting on this.

**Comment by CharlesOkwuagwu** [comment]
Hi,

You could require the HKDF constructor to take a version variable then, set iterationStartOffset to either 1 or 0 depending on the version... for  a start

I think the HKDF for V3 is more in line with the standard... or is it the other way around?

**Comment by CharlesOkwuagwu** [comment]
IT would be good to support both versions just in case

**Comment by joebandenburg** [comment]
@CharlesOkwuagwu Absolutely, I've put a few "TODO" comments in the code to indicate where changes would need to be made to support version 2. I don't think it would be too difficult. The biggest unknown is if both the node.js [crypto](http://nodejs.org/api/crypto.html) module and [WebCrypto](http://www.w3.org/TR/WebCryptoAPI/) support AES-256-CTR with a custom CTR (see [libaxolotl-android](https://github.com/WhisperSystems/libaxolotl-android/blob/master/java/src/main/java/org/whispersystems/libaxolotl/SessionCipher.java#L433)). They both support counter mode, but it's not entirely clear if node.js supports initialising the counter to a specific value.

**Comment by CharlesOkwuagwu** [comment]
@joebandenburg Interesting. I've been toying with porting Axolotl to .net, the full libaxolotl-android seems like over-kill compared to the Web-browser version or, better yet your version. There is no  CTR mode in .net at all... that would make V2 support tricky, without a 3rd party lib.

Quick question, you leave Store and Crypto as service interfaces in your implementation.
Where is the concrete implementation for these?

(Over-kill WRT https://github.com/trevp/axolotl/wiki, i've actually come a cross a surprisingly simple python implementation, here on github)

**Comment by joebandenburg** [comment]
> I think the HKDF for V3 is more in line with the standard... or is it the other way around?

Yes, from my reading of the [spec](https://tools.ietf.org/html/rfc5869) starting with 1 (version 3) is in line with the standard.

> I've been toying with porting Axolotl to .net

Awesome :) I'd be very interested to hear your feedback about my code. I've tried to make the implementation as clear as possible to help others understand it. I found Axolotl quite complex initially.

> without a 3rd party lib.

You may need a 3rd party library anyway to support Curve25519 and Ed25519, which are required by both versions of the protocol. (They are needed for the `calculateAgreement` and `sign` methods on the `Crypto` interface.)

> Quick question, you leave Store and Crypto as service interfaces in your implementation.
> Where is the concrete implementation for these?

There aren't any implementations yet. You can see a sort-of node.js implementation in `test/unit/FakeCrypto.js`, although take note that the `calculateAgreement` and `sign` methods are not suitable for a real implementation. I've been pondering including an implementation of Curve25519/Ed25519 in libaxolotl-javascript, so the user doesn't need to provide it themselves.

I will be adding documentation to the README to explain the exact requirements for each method of `Crypto` interface.

**Comment by joebandenburg** [comment]
> (Over-kill WRT https://github.com/trevp/axolotl/wiki, i've actually come a cross a surprisingly simple python implementation, here on github)

My implementation, like libaxolotl-android, is specific to TextSecure and therefore has the same "non-standard" bits (WTR https://github.com/trevp/axolotl/wiki).

**Comment by CharlesOkwuagwu** [comment]
Please can you share more Test vectors.
Validating one's implementation seems to be quite difficult without comparing code accuracy to actual expected behaviour (test vectors)

**Comment by CharlesOkwuagwu** [comment]
@ "You may need a 3rd party library anyway to support Curve25519 and Ed25519, which are required by both versions of the protocol. (They are needed for the calculateAgreement and sign methods on the Crypto interface.)"

This part I have successfully completed. surprisingly the Crypto and HKDF were the easier parts to get through.

Lookup CodesInChaos, he has a sweet C# implementation of NaCL which covers Ed25519 and has all the building blocks for Curve25519. 

Building on that I could easily reproduce Trevor Perrins  conversions from Curve25519 to ed25519 here https://github.com/WhisperSystems/libaxolotl-android/blob/master/jni/ed25519/additions/curve_sigs.c 

Maybe I'll post that as a GitGist

**Comment by CharlesOkwuagwu** [comment]
It's getting the "Ratcheting Waltz" implemented correctly that's the tricky bit :(

Your code and documentation really helps a lot.

Thanks!
